File: src/karabo/core/ComputeDevice.cc
/* 
 * $Id$
 * 
 * Authors: <burkhard.heisen@xfel.eu>
 *          <steffen.hauf@xfel.eu>
 * 
 * Created on July 22, 2013, 9:10 AM
 * 
 * Copyright (C) European XFEL GmbH Hamburg. All rights reserved.
 */

#include "ComputeDevice.hh"
#include <karabo/xms/SlotElement.hh>

using namespace std;
using namespace karabo::util;
using namespace karabo::core;
using namespace karabo::net;
using namespace karabo::io;
using namespace karabo::xms;
using namespace karabo::xip;

namespace karabo {
    namespace core {


        void ComputeDevice::expectedParameters(Schema& expected) {

            SLOT_ELEMENT(expected).key("start")
                    .displayedName("Compute")
                    .description("Starts computing if data is available.")
                    .allowedStates("Ok.Paused Ok.Finished Ok.Compute Ok.Ready")
                    .commit();

            SLOT_ELEMENT(expected).key("pause")
                    .displayedName("Pause")
                    .description("Will finish current computation and pause. End of stream events are not forwarded in the paused state if expert-mode parameter setEOSPauseAllowed is false!")
                    .allowedStates("Ok.Computing")
                    .commit();

            SLOT_ELEMENT(expected).key("abort")
                    .displayedName("Abort")
                    .description("Try to abort the current computation by interrupting the compute thread and emptying the current buffer. "
                    "The compute thread is afterwards restarted and resetting the device enables further computations."
                    "Usage Scenario: abort a long running computation which has gone wrong.")
                    .allowedStates("Ok.Computing Ok.Paused")
                    .commit();

            SLOT_ELEMENT(expected).key("endOfStream")
                    .displayedName("End-Of-Stream")
                    .description("Send an End-Of-Stream Token")
                    .allowedStates("Ok.Paused Ok.Computing Ok.Finished")
                    .commit();

            SLOT_ELEMENT(expected).key("reset")
                    .displayedName("Reset")
                    .description("Completely reset this device")
                    .allowedStates("Error.Ready Ok.Aborted Ok")
                    .commit();

            BOOL_ELEMENT(expected).key("autoCompute")
                    .displayedName("Auto Compute")
                    .description("Trigger computation automatically once data is available. The device will stay in the computing state until manually paused or aborted.")
                    .reconfigurable()
                    .assignmentOptional().defaultValue(true)
                    .commit();

            BOOL_ELEMENT(expected).key("autoEndOfStream")
                    .displayedName("Auto end-of-stream")
                    .description("If true, automatically forwards the end-of-stream signal to all connected (downstream) devices. Requires the device to not be paused if autoCompute is enabled.")
                    .reconfigurable()
                    .expertAccess()
                    .assignmentOptional().defaultValue(true)
                    .commit();

            BOOL_ELEMENT(expected).key("autoUpdate")
                    .displayedName("Auto update")
                    .description("If true, automatically updates all input and output channels after compute was called")
                    .reconfigurable()
                    .expertAccess()
                    .assignmentOptional().defaultValue(true)
                    .commit();

            BOOL_ELEMENT(expected).key("autoIterate")
                    .displayedName("Auto iterate")
                    .description("If true, automatically iterates cyclic workflows")
                    .reconfigurable()
                    .expertAccess()
                    .assignmentOptional().defaultValue(true)
                    .commit();
            
            BOOL_ELEMENT(expected).key("pauseEOSAllowed")
                    .displayedName("Forward end-of-stream when paused")
                    .description("If true, end-of-stream events are forwarded when the device is paused. If false, they are only allowed in the computing and finished states.")
                    .reconfigurable()
                    .expertAccess()
                    .assignmentOptional().defaultValue(true)
                    .commit();

            INT32_ELEMENT(expected).key("iteration")
                    .displayedName("Iteration")
                    .description("The current iteration")
                    .readOnly()
                    .initialValue(0)
                    .commit();
            
            UINT32_ELEMENT(expected).key("abortTimeOut")
                    .displayedName("Abort timeout (ms)")
                    .description("Time out to wait for compute to finish before calling interrupt to choke it off. The interrupt call will have the same timeout.")
                    .assignmentOptional().defaultValue(5000)
                    .reconfigurable()
                    .expertAccess()
                    .commit();
            
        }


        ComputeDevice::ComputeDevice(const Hash& input) : Device<>(input), m_isAborted(false), m_isPaused(true), m_isEndOfStream(false), m_deviceIsDead(false), m_nEndOfStreams(0), m_iterationCount(0) {

            m_computeThread = boost::thread(boost::bind(&karabo::core::ComputeDevice::doCompute, this));

            SLOT0(start);
            SLOT0(pause);
            SLOT0(abort);
            SLOT0(endOfStream);
            SLOT0(reset);


        }


        ComputeDevice::~ComputeDevice() {
            setDeviceDead();
            
            //notify compute thread to acknowledge death of device
            m_computeCond.notify_one();

            if (m_computeThread.joinable()) m_computeThread.join();

            KARABO_LOG_DEBUG << "dead.";
        }


        void ComputeDevice::setDeviceDead() {
       
            m_deviceIsDead.store(true, boost::memory_order_relaxed);
        }


        void ComputeDevice::_onInputAvailable(const karabo::io::AbstractInput::Pointer&) {
            
            
            //If we don't want to auto compute, do nothing
            if(get<bool>("autoCompute")){
                if(!m_isPaused) {
                    m_computeCond.notify_one();
                }  
            } 
            
        }


        void ComputeDevice::_onEndOfStream() {

            size_t m_expectedEndOfStreams = 0;

            // Count all channels that should respond to eos
            const InputChannels& inputChannels = this->getInputChannels();
            for (InputChannels::const_iterator it = inputChannels.begin(); it != inputChannels.end(); ++it) {
                if (it->second->respondsToEndOfStream()) m_expectedEndOfStreams++;
            }

            m_nEndOfStreams++; // Counts this function call
            if (m_nEndOfStreams >= m_expectedEndOfStreams) {
                m_nEndOfStreams = 0;
                m_isEndOfStream = true;
                this->endOfStream();
            }
        }


        void ComputeDevice::endOfStreamAction() {

            this->onEndOfStream();
            if (get<bool>("autoEndOfStream")) {
                const OutputChannels& outputChannels = this->getOutputChannels();
                for (OutputChannels::const_iterator it = outputChannels.begin(); it != outputChannels.end(); ++it)
                    it->second->signalEndOfStream();
            }
        }

        // User hook
        void ComputeDevice::onEndOfStream() {

        }


        void ComputeDevice::update() {
            if (get<bool>("autoUpdate")) {
                const InputChannels& inputChannels = this->getInputChannels();
                const OutputChannels& outputChannels = this->getOutputChannels();
                for (InputChannels::const_iterator it = inputChannels.begin(); it != inputChannels.end(); ++it)
                    it->second->update();
                for (OutputChannels::const_iterator it = outputChannels.begin(); it != outputChannels.end(); ++it)
                    it->second->update();
            }
        }


        bool ComputeDevice::isAborted() const {
            return m_isAborted.load(boost::memory_order_consume);
        }


        void ComputeDevice::connectAction() {
            KARABO_LOG_FRAMEWORK_DEBUG << "Connecting "<<this->getInputChannels().size()<<" IO channels";
            this->connectInputChannels();
            this->set<bool>("autoCompute", this->checkAutoComputeValidity(this->get<bool>("autoCompute")));
           
        }
        
        void ComputeDevice::readyStateOnEntry(){
           
            if(this->get<bool>("autoCompute")){
                m_isPaused = false;
                this->start();
            } else {
                this->pause();
            }
        }
        
        void ComputeDevice::preReconfigure(karabo::util::Hash& incomingReconfiguration){
            
            incomingReconfiguration.set<bool>("autoCompute", this->checkAutoComputeValidity(incomingReconfiguration.get<bool>("autoCompute")));
            
        }

        bool ComputeDevice::checkAutoComputeValidity(const bool & requestedValue){
            //if now input channels are connected autoCompute does not make any sense
            if(this->getInputChannels().size() == 0 &&  requestedValue){
                KARABO_LOG_WARN<<"This device does not have any input channels connected. Setting autoCompute=false";
                return false;
            }
            return requestedValue; 
        }

        

        bool ComputeDevice::canCompute() {

            //            size_t m_expectedEndOfStreams = 0;
            //
            //            const InputChannels& inputChannels = this->getInputChannels();
            //            for (InputChannels::const_iterator it = inputChannels.begin(); it != inputChannels.end(); ++it) {
            //                if (it->second->respondsToEndOfStream()) m_expectedEndOfStreams++;
            //            }


            //            for (InputChannels::const_iterator it = inputChannels.begin(); it != inputChannels.end(); ++it) {
            //                if (!it->second->canCompute() && (it->second->respondsToEndOfStream() || m_expectedEndOfStreams == 0)) return false;
            //            }

            // Moved logic of the upper code into the canCompute function of the input channel, thus:
            const InputChannels& inputChannels = this->getInputChannels();
            for (InputChannels::const_iterator it = inputChannels.begin(); it != inputChannels.end(); ++it) {
                if (!it->second->canCompute()) {
                    return false;
                }
            }
            return true;
        }


        void ComputeDevice::computingStateOnEntry() {        
            m_computeCond.notify_one(); 
        }


        void ComputeDevice::doCompute() try {

            boost::unique_lock<boost::mutex> lock(m_computeMutex);
            
            while (true) {
                
                //the condition variable here will block until
                // a) an input becomes available and triggers it
                // b) the device goes back into the compute state from one of the other states
                // c) the device is killed, i.e. from the destructor, so that the thread returns
                {
                   m_computeCond.wait(lock);
                }
                
                
                
                //unset promise: it will be set again when this iteration has finished.
                //an unset promise blocks any transitions to pause if a computation is still in process
                boost::promise<bool> p(m_workIsFinished.move());
        
                if (m_deviceIsDead.load(boost::memory_order_consume)) {
                    m_workIsFinished.set_value(true);
                    return;
                }

                if (!m_isAborted.load(boost::memory_order_consume) && this->canCompute()) {
                    
                    try {
                        
                      this->compute();
                      
                    } catch (const karabo::util::Exception& e) {
                        
                        KARABO_LOG_ERROR << "Caught exception in compute thread: " << e.userFriendlyMsg();
                        
                    } catch (const boost::thread_interrupted& ){
                        
                        KARABO_LOG_WARN << "Compute thread has been interrupted by call to abort.";
                        return;
                        
                    } catch (...) {
                        
                        KARABO_LOG_ERROR << "Caught unknown exception in compute thread";
                        
                    }
                    
                    //if we are not in autoCompute mode this compute call is now finished, and we change to finished state
                    //otherwise, the loop will return to the condition variable, which is triggered by the inputs
                    if(!this->get<bool>("autoCompute")){
                        this->computeFinished();
                    }

                }
                
                try {
                    
                    this->update();
                    
                } catch (const karabo::util::Exception& e) {
                    
                      KARABO_LOG_ERROR << "Caught exception while updating channels: " << e.userFriendlyMsg();
                      
                } catch (...) {
                    
                      KARABO_LOG_ERROR << "Caught unknown exception while updating channels";
                }
                
                m_workIsFinished.set_value(true);
               
            }
        } catch (const karabo::util::Exception& e) {
            KARABO_LOG_ERROR << e;
        }


        void ComputeDevice::computingStateOnExit() {
            
        }

        void ComputeDevice::setComputationAborted(){
            this->abort();
        }
       
        bool ComputeDevice::registerAbort() {
          
            m_isAborted.store(true, boost::memory_order_release);
          
            unsigned int timeout = this->get<unsigned int>("abortTimeOut");
           
            
            boost::unique_future<bool> f = m_workIsFinished.get_future();
            boost::future_status status = f.wait_for(boost::chrono::milliseconds(timeout));
            if(status == boost::future_status::timeout){
                //we kill and restart the computing thread if the wait timed out. In case it succeed we can keep the thread alive.
           
                //the cleanest way is if interruption points are set
                m_computeThread.interrupt();
                if(m_computeThread.try_join_for(boost::chrono::milliseconds(timeout))) {
                    m_computeThread = boost::thread(boost::bind(&karabo::core::ComputeDevice::doCompute, this));
                } else {
                    KARABO_LOG_WARN<<"Thread could not be properly interrupted.  Consider using boost::interruption_point() within your compute implementation.";
                    
                    //The following part would work, but is not safe, as terminating the thread has a high likelihood of tearing the complete device (and device server) down
                    
                    /* int killSucceeded = 0;
                    try{
#if defined(_MSC_VER)
                        killSucceeded = !(bool)(TerminateThread(m_computeThread.native_handle()));
#else
                        killSucceeded = pthread_kill(m_computeThread.native_handle(), 0);
#endif
                    } catch (...){
                        killSucceeded = 1;
                    }
                    if(killSucceeded != 0){
                        KARABO_LOG_WARN<<"Could not terminate thread. Detaching it instead and hoping that it dies gracefully...";
                        try{
                            m_computeThread.detach();
                        } catch (...){
                            KARABO_LOG_WARN<<"Could not detach thread!";
                        }
                    }*/
                }
                
            }
            
            //clear any data in the buffer which was being worked upon
            this->update();
   
            return true;
        }


        bool ComputeDevice::registerPause() {
            //wait for any work to finish
            boost::unique_future<bool> f = m_workIsFinished.get_future();
            f.wait();
            
            return true;
        }
        
         void ComputeDevice::pausedStateOnEntry(){
            m_isPaused = true;
        }
        
        void ComputeDevice::pausedStateOnExit(){
            m_isPaused = false;
            
        }

        bool ComputeDevice::checkPauseEOSAllowed(){
            return this->get<bool>("pauseEOSAllowed");
        }
        
        void ComputeDevice::finishedOnEntry() {
            m_iterationCount++;
            m_isEndOfStream = false;
            
        }


        void ComputeDevice::finishedOnExit() {
            set("iteration", m_iterationCount);
            
        }

        void ComputeDevice::abortedOnEntry() {
            
        }

        void ComputeDevice::abortedOnExit() {
           
            m_isAborted.store(false, boost::memory_order_release);
        }


        int ComputeDevice::getCurrentIteration() const {
            return get<int>("iteration");
        }
    }
}

File: src/pythonKarabo/karabo_package/compute_device.py
# To change this template, choose Tools | Templates
# and open the template in the editor.

__author__="Sergey Esenov <serguei.essenov at xfel.eu>"
__date__ ="$Sep 10, 2013 4:17:06 PM$"

from abc import abstractmethod
import threading

import karabo.base_fsm as base
from karabo.configurator import *
from karabo.device import *
from karabo.fsm import *


@KARABO_CLASSINFO("ComputeFsm", "1.0")
class ComputeFsm(base.BaseFsm):
    
    @staticmethod
    def expectedParameters(expected):
        (
        SLOT_ELEMENT(expected).key("start")
                .displayedName("Compute")
                .description("Starts computing if data is available")
                .allowedStates("Ok:Ready")
                .commit()
                ,
        SLOT_ELEMENT(expected).key("pause")
                .displayedName("Pause")
                .description("Will finish current computation and pause")
                .allowedStates("Ok:Ready")
                .commit()
                ,
        SLOT_ELEMENT(expected).key("abort")
                .displayedName("Abort")
                .description("Abort contribution to this run, fully disconnect")
                .allowedStates("Ok:Ready Ok:Computing Ok:WaitingIO Ok:Paused")
                .commit()
                ,
        SLOT_ELEMENT(expected).key("endOfStream")
                .displayedName("End-Of-Stream")
                .description("Completely reset this device")
                .allowedStates("Ok:Ready")
                .commit()
                ,
        SLOT_ELEMENT(expected).key("reset")
                .displayedName("Reset")
                .description("Completely reset this device")
                .allowedStates("Error:Ready Error:Computing Error:WaitingIO Ok:Finished Ok:Aborted")
                .commit()
                ,
        )
        
    def __init__(self, configuration):
        super(ComputeFsm, self).__init__(configuration)
        
        #**************************************************************
        #*                        Events                              *
        #**************************************************************
        KARABO_FSM_EVENT(self, 'ErrorFoundEvent',       'errorFound')
        KARABO_FSM_EVENT(self, 'ResetEvent',            'reset')
        KARABO_FSM_EVENT(self, "StartEvent",            'start')
        KARABO_FSM_EVENT(self, "EndOfStreamEvent",      'endOfStream')
        KARABO_FSM_EVENT(self, 'PauseEvent',            'pause')
        KARABO_FSM_EVENT(self, 'AbortEvent',            'abort')
        KARABO_FSM_EVENT(self, 'ComputeFinishedEvent',  'computeFinished')
        KARABO_FSM_EVENT(self, 'UpdatedIOEvent',        'updatedIO')

        #**************************************************************
        #*                        States                              *
        #**************************************************************
        KARABO_FSM_STATE('Ok')
        KARABO_FSM_INTERRUPT_STATE('Error', 'ResetEvent')
        KARABO_FSM_STATE('ConnectingIO')
        KARABO_FSM_STATE('Ready',     self.readyStateOnEntry)
        KARABO_FSM_STATE('Computing', self.computingStateOnEntry, self.computingStateOnExit)
        KARABO_FSM_STATE('WaitingIO', self.waitingIOOnEntry,      self.waitingIOOnExit)
        KARABO_FSM_STATE('Paused')
        KARABO_FSM_STATE('Finished',  self.finishedOnEntry,       self.finishedOnExit)
        KARABO_FSM_STATE('Aborted',   self.abortedOnEntry)

        #**************************************************************
        #*                    Transition Actions                      *
        #**************************************************************
        #KARABO_FSM_NO_TRANSITION_ACTION(self.noStateTransition)  # <- this is defined in BaseFsm
        KARABO_FSM_ACTION('ErrorFoundAction',    self.errorFoundAction, str, str)
        KARABO_FSM_ACTION('ConnectAction',       self.connectAction)
        KARABO_FSM_ACTION('EndOfStreamAction',   self.endOfStreamAction)
        KARABO_FSM_ACTION('NextIterationAction', self.onNextIteration)

        #**************************************************************
        #*                    Guards                                  *
        #**************************************************************
        KARABO_FSM_GUARD('CanCompute', self.canCompute)
        KARABO_FSM_GUARD('AbortGuard', self.registerAbort)
        KARABO_FSM_GUARD('PauseGuard', self.registerPause)
        
        #**************************************************************
        #*                    State Machine                           *
        #**************************************************************
        onSTT = [
        # Source-State        Event              Target-State    Action             Guard
            ('ConnectingIO', 'none',             'Ready',     'ConnectAction',      'none'),
            ('Ready',        'StartEvent',       'Computing', 'none',               'CanCompute'),
            ('Ready',        'PauseEvent',       'Paused',    'none',               'none'),
            ('Ready',        'AbortEvent',       'Aborted',   'none',               'none'),
            ('Ready',        'EndOfStreamEvent', 'Finished',  'EndOfStreamAction',  'none'),
            ('Paused',       'ResetEvent',       'Ready',     'none',               'CanCompute'),
            ('Computing','ComputeFinishedEvent', 'WaitingIO', 'none',               'none'),
            ('Computing',    'AbortEvent',       'Aborted',   'none',               'AbortGuard'),
            ('Computing',    'PauseEvent',       'Paused',    'none',               'PauseGuard'),
            ('WaitingIO',    'UpdatedIOEvent',   'Ready',     'none',               'none'),
            ('Aborted',      'ResetEvent',       'Ready',     'none',               'none'),
            ('Finished',     'ResetEvent',       'Ready',     'none',               'none'),
            ('Finished',     'StartEvent',       'Computing', 'NextIterationAction','none'),
            ('Ok',           'ErrorFoundEvent',  'Error',     'ErrorFoundAction',   'none'),
            ('Error',        'ResetEvent',       'Ok',        'none',               'none')
        ]

        #                        Name     Transition-Table   Initial-State
        KARABO_FSM_STATE_MACHINE('StateMachine', onSTT, ('Ok', 'ConnectingIO'))
        self.fsm = KARABO_FSM_CREATE_MACHINE('StateMachine')
    
    def getFsm(self):
        return self.fsm

    ########################################################
    #  Guards, Transition Actions, State Machine hooks...  #
    ########################################################
    
    @abstractmethod
    def readyStateOnEntry(self):
        pass
    
    @abstractmethod
    def computingStateOnEntry(self):
        pass
    
    @abstractmethod
    def computingStateOnExit(self):
        pass
    
    @abstractmethod
    def waitingIOOnEntry(self):
        pass
    
    @abstractmethod
    def waitingIOOnExit(self):
        pass
    
    @abstractmethod
    def finishedOnEntry(self):
        pass
    
    @abstractmethod
    def finishedOnExit(self):
        pass
    
    @abstractmethod
    def abortedOnEntry(self):
        pass
    
    @abstractmethod
    def errorFoundAction(self, s1, s2):
        pass
    
    @abstractmethod
    def connectAction(self):
        pass
    
    @abstractmethod
    def endOfStreamAction(self):
        pass
    
    @abstractmethod
    def onNextIteration(self):
        pass
    
    @abstractmethod
    def canCompute(self):
        return True
    
    @abstractmethod
    def registerAbort(self):
        return True
    
    @abstractmethod
    def registerPause(self):
        return True


@KARABO_CLASSINFO("PythonComputeDevice", "1.0")
class PythonComputeDevice(PythonDevice, ComputeFsm):
    
    def __init__(self, configuration):
        super(PythonComputeDevice, self).__init__(configuration)
        self.isAborted = False
        self.isEndOfStream = False
        self.deviceIsDead = False
        self.nEndOfStreams = 0
        self.iterationCount = 0
        self.computeLock = threading.Lock()
        self.computeLock.acquire(True)
        self.computeThread = threading.Thread(target=self.doCompute)
        self.computeThread.start()
        self.waitingIOLock = threading.Lock()
        self.waitingIOLock.acquire(True)
        self.waitingIOThread = threading.Thread(target=self.doWait)
        self.waitingIOThread.start()

    def __del__(self):
        ''' PythonComputeDevice destructor '''
        self.deviceIsDead = True
        self.computeLock.release()
        self.computeThread.join()
        self.waitingIOLock.release()
        self.waitingIOThread.join()
        super(PythonComputeDevice, self).__del__()
        
    @staticmethod
    def expectedParameters(expected):
        (
        BOOL_ELEMENT(expected).key("autoCompute")
                .displayedName("Auto Compute")
                .description("Trigger computation automatically once data is available")
                .reconfigurable()
                .assignmentOptional().defaultValue(True)
                .commit()
                ,
        BOOL_ELEMENT(expected).key("autoEndOfStream")
                .displayedName("Auto end-of-stream")
                .description("If true, automatically forwards the end-of-stream signal to all connected (downstream) devices")
                .reconfigurable()
                .assignmentOptional().defaultValue(True)
                .commit()
                ,
        BOOL_ELEMENT(expected).key("autoIterate")
                .displayedName("Auto iterate")
                .description("If true, automatically iterates cyclic workflows")
                .reconfigurable()
                .assignmentOptional().defaultValue(True)
                .commit()
                ,
        INT32_ELEMENT(expected).key("iteration")
                .displayedName("Iteration")
                .description("The current iteration")
                .readOnly()
                .initialValue(0)
                .commit()
                ,
        )
        
    def KARABO_INPUT_CHANNEL(self, type, name, configuration):
        if type.__name__ == "InputHash":
            return self._ss.createInputChannelHash(name, configuration, self._onInputAvailable, self._onEndOfStream)
    
    def KARABO_OUTPUT_CHANNEL(self, type, name, configuration):
        if type.__name__ == "OutputHash":
            return self._ss.createOutputChannelHash(name, configuration)
        
    @abstractmethod    
    def compute(self):
        ''' Put your specific algorithms here '''

    def _onInputAvailable(self, input):
        if self.get("state") == "Ok.Finished" and not self.get("autoIterate"):
            pass
        else:
            self.start()
    
    def _onEndOfStream(self):
        self.nEndOfStreams = self.nEndOfStreams + 1
        if self.nEndOfStreams >= len(self._ss.getInputChannels()):
            self.nEndOfStreams = 0
            self.isEndOfStream = True
            if self.get("autoEndOfStream"): 
                self.endOfStream();
    
    def endOfStreamAction(self):
        self.onEndOfStream()
        outputChannels = self._ss.getOutputChannels()
        for c in outputChannels:
            #channel.signalEndOfStream()
            c.data().signalEndOfStream()
            
    
    def onEndOfStream(self):
        ''' Override this function for specializing the endOfStream behavior '''
        
    def  update(self):
        ''' Override this function for specializing the update behaviors of your IO channels '''
        inputChannels = self._ss.getInputChannels()
        outputChannels = self._ss.getOutputChannels()
        for c in inputChannels:
            c.data().update()
        for c in outputChannels:
            c.data().update()
        
    def isAborted(self):
        return self.isAborted;
    
    def connectAction(self):
        return self._ss.connectInputChannels()
    
    def readyStateOnEntry(self):
        if self.isEndOfStream and self.get("autoEndOfStream"):
            self.endOfStream()
        elif self.isAborted:
            self.abort()
        elif len(self._ss.getInputChannels()) > 0 and self.get("autoCompute"):
            self.start()
            
    def canCompute(self):
        inputChannels = self._ss.getInputChannels()
        for c in inputChannels:
            if not c.data().canCompute():
                return False
        return True
    
    def computingStateOnEntry(self):
        self.computeLock.release()
        
    def doCompute(self):
        while True:
            self.computeLock.acquire(True)
            if self.deviceIsDead:
                return
            self.compute()
            self.computeLock.release()
            if not self.isAborted:
                self.computeFinished()
        
    def computingStateOnExit(self):
        self.computeLock.acquire(True)
        
    def registerAbort(self):
        self.isAborted = True;
        return True
    
    def registerPause(self):
        return True
    
    def waitingIOOnEntry(self):
        self.waitingIOLock.release()

    def doWait(self):
        while True:
            self.waitingIOLock.acquire(True)
            if self.deviceIsDead:
                return
            try:
                self.update()
            except Exception as e:
                self.waitingIOLock.release()
                self.errorFound("Exception caught", str(e))
                return
            self.waitingIOLock.release()
            self.updatedIO()
            
    def waitingIOOnExit(self):
        self.waitingIOLock.acquire(True)
        
    def finishedOnEntry(self):
        self.iterationCount = self.iterationCount + 1
        self.isEndOfStream = False
        
    def finishedOnExit(self):
        self.set("iteration", self.iterationCount)
        
    def abortedOnEntry(self):
        self.isAborted = False
        
